Mslm SDK

Harisabdullah opened the pull request.

UmanShahzad commented on README.md:
````diff
@@ -0,0 +1,59 @@
+
+```python
+from mslm import Mslm
+from otp.otp_req import OtpReq
````
We shouldn't have to import from a root `otp` package for this. This could easily lead to name conflicts with other packages with such root names. Rather it can be nested.

We also need to think about how this links to `pip install` and being able to install individual packages only instead of the whole thing by force, e.g. `pip install mslm` for the whole thing, but `pip install mslm-otp` for OTP an `pip install mslm-email-verify` for email verify, etc.

UmanShahzad commented on README.md:
```diff
@@ -0,0 +1,59 @@
+
+## About Mslm
+
+mslm focuses on producing world-class business solutions. It’s the
```
Capital M

UmanShahzad commented on lib/lib.py:
```diff
@@ -0,0 +1,138 @@
+        - req_and_resp(self, t_url, opt, method='GET', data=None): Makes an HTTP request and returns the response.
+    """
+
+    gson = None  # Assuming you have a Gson-like library in Python
```
We can remove no?

UmanShahzad commented on lib/lib.py:
```diff
@@ -0,0 +1,138 @@
+import json
+from typing import Optional
+from urllib.parse import urlparse
+from requests import Session
```
We're assuming this library exists but have no `requirements.txt` or any `setup.py` or anything.

See github.com/ipinfo/python for some ideas

Harisabdullah replied:
Added `requirements.txt`

UmanShahzad commented on otp/otp_req_opts.py:
```diff
@@ -0,0 +1,81 @@
+            self.opts.req_opts = req_opts
+            return self
+
+        def with_disable_url_encode(self, disable_url_encode: bool) -> 'OtpReqOpts.Builder':
```
Did we have this option in Golang too for OTP or just for sv?

Harisabdullah commented:
@UmanShahzad 
Salam brother
- added `requirements.txt`
- changes around getting it ready for `pip` and "`import structure`".
- implemented error handling
Also
- kept the `_` rather than `-` for package directories according to community guidelines. hence the `pip install .\mslm_email_verify` rather than `pip install .\mslm-email-verify` .


Please let me know if changes are required

UmanShahzad reviewed (commented):
I think we've made progress but it's not there yet, the structure to make it ready for being installed via `pypi` and `pip install` may need to be changed more.

Or maybe I'm wrong - what happens when you try to setup a distribution package (see scripts in II python repo) with one of these `setup.py` while sitting in its directory? What files does it try to output inside of the zipped file?

If that doesn't go well or produces weird results, I'd read this stuff:

- https://packaging.python.org/en/latest/flow/
- https://packaging.python.org/en/latest/tutorials/packaging-projects/
- https://stackoverflow.com/questions/75408641/whats-the-difference-between-the-tool-poetry-and-project-tables-in-pyprojec
- https://sinoroc.gitlab.io/kb/python/packaging_tools_comparisons.html#build-back-ends

We could go in the direction of using a `pyproject` at the root, with the `setuptools` build-backend. Not sure if it helps.

Harisabdullah commented:
@UmanShahzad 
I tried running `python setup.py sdist bdist_wheel` and it generates `Gzipped Tar Archive` and `Wheel Archive`.
Now we need a `pypi` account to run `twine upload dist/*` and it would upload to `pypi`.

UmanShahzad commented:
Sure but what is in it? Show file structure within the package

Harisabdullah commented:
> Sure but what is in it? Show file structure within the package

The following is file structure after running `python setup.py sdist bdist_wheel` for `mslm_email_verify`
- build/
  - bdist.linux-x86_64/
  - lib/
    - lib/
      - __init__.py
      - lib.py
      - mslm_errors.py
      - req_opts.py
    - mslm_email_verify/
      - email_verify.py
      - __init__.py
      - single_verify_resp_mx.py
      - single_verify_resp.py
- dist/
  - mslm_email_verify-1.0.0-py3-none-any.whl
  - mslm-email-verify-1.0.0.tar.gz
- email_verify.py
- __init__.py
- lib/
  - __init__.py
  - lib.py
  - mslm_errors.py
  - req_opts.py
- mslm_email_verify/
  - email_verify.py
  - __init__.py
  - single_verify_resp_mx.py
  - single_verify_resp.py
- mslm_email_verify.egg-info/
  - dependency_links.txt
  - PKG-INFO
  - requires.txt
  - SOURCES.txt
  - top_level.txt
- setup.py
- single_verify_resp_mx.py
- single_verify_resp.py
gle_verify_resp.py

UmanShahzad commented:
@Harisabdullah what's in the `.tar.gz` files, that's the thing that IIRC will be uploaded.

Harisabdullah commented:
- mslm-1.0.0/
  - LICENSE
  - PKG-INFO
  - README.md
  - lib/
    - __init__.py
    - lib.py
    - mslm_errors.py
    - req_opts.py
  - mslm/
    - __init__.py
    - mslm.py
  - mslm.egg-info/
    - PKG-INFO
    - SOURCES.txt
    - dependency_links.txt
    - requires.txt
    - top_level.txt
  - mslm_email_verify/
    - __init__.py
    - email_verify.py
    - single_verify_resp.py
    - single_verify_resp_mx.py
  - mslm_otp/
    - __init__.py
    - otp.py
    - otp_resp.py
    - otp_send_req.py
    - otp_token_verify_req.py
  - setup.cfg
  - setup_mslm.py

Harisabdullah commented:
- mslm-email-verify-1.0.0/
  - LICENSE
  - PKG-INFO
  - README.md
  - lib/
    - __init__.py
    - lib.py
    - mslm_errors.py
    - req_opts.py
  - mslm_email_verify/
    - __init__.py
    - email_verify.py
    - single_verify_resp.py
    - single_verify_resp_mx.py
  - mslm_email_verify.egg-info/
    - PKG-INFO
    - SOURCES.txt
    - dependency_links.txt
    - requires.txt
    - top_level.txt
  - setup.cfg
  - setup_mslm_email_verify.py

Harisabdullah commented:
- mslm-otp-1.0.0/
  - LICENSE
  - PKG-INFO
  - README.md
  - lib/
    - __init__.py
    - lib.py
    - mslm_errors.py
    - req_opts.py
  - mslm_otp/
    - __init__.py
    - otp.py
    - otp_resp.py
    - otp_send_req.py
    - otp_token_verify_req.py
  - mslm_otp.egg-info/
    - PKG-INFO
    - SOURCES.txt
    - dependency_links.txt
    - requires.txt
    - top_level.txt
  - setup.cfg
  - setup_mslm_otp.py

UmanShahzad reviewed (commented):
Alright let's proceed in two ways:

- Please rename `lib` to `mslm_lib`. Otherwise I think (but am not 100% sure) that `lib` will suddenly become a valid absolute import path for people who install our libraries, and that could lead to some issues. In fact, I'm not clear on what even happens if someone installs two of our libraries (e.g. mslm-otp and mslm-email-verify), and they try to do `import mslm_lib` - is this even possible?
- We need to do a proper end-to-end test of the installation and packaging process it seems, otherwise this'll be hard to know for sure. You can go ahead and register on Pypi and do the `twine`-based upload after getting a valid API key and what not. You can name the packages `mslm`, `mslm-otp`, `mslm-email-verify` and so on. When done and uploaded, you can give it a spin yourself and see how it goes - try installing multiple of those packages separately into a test python package, and see what happens and if it causes issues when you try to do imports and what not.

UmanShahzad commented on README.md:
Please fix the README

UmanShahzad commented on CONTRIBUTING.md:
?

UmanShahzad commented on scripts/publish.sh:
```diff
@@ -0,0 +1,17 @@
+echo "Cleaning up..."
+python3 setup.py clean --all
+
+# clean egs
+rm -rf mslm.egg-info
+rm -rf mslm_email_verify.egg-info
+rm -rf mslm_otp.egg-info
```
```suggestion
# clean build artifacts
rm -rf *.egg-info
```

UmanShahzad commented on scripts/publish.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+echo "Building mslm..."
```
```suggestion
echo "Building..."
```

UmanShahzad commented on scripts/publish.sh:
```diff
@@ -0,0 +1,17 @@
+echo "Building mslm..."
+python3 setup.py sdist bdist_wheel
+
+echo "Uploading ..."
```
```suggestion
echo "Uploading..."
```

UmanShahzad commented on scripts/publish.sh:
Dont we need this script now in each subfolder, i.e. a copy? Or maybe you can have a common script here and then one inside of each that will just call this script with correct pathing?

UmanShahzad commented on setup.py:
Don't we need this as a copy for each library if we are gonna be having different packages uploaded?

UmanShahzad reviewed (commented):
You said you tested end-to-end but I don't see any test package on pypi?

Harisabdullah commented:
> You said you tested end-to-end but I don't see any test package on pypi?
@UmanShahzad 
[Mslm test sdk on pypi](https://pypi.org/project/mslm-test-haris-m)

UmanShahzad commented:
Coolio!

Harisabdullah commented:
@UmanShahzad 
I have created seperate `setup.py`s for each package. I have tested them, the following are live test packages.
[Mslm](https://pypi.org/project/mslm-test2/)
[Mslm-otp](https://pypi.org/project/mslm-otp-test2/)
[Mslm-email-verify](https://pypi.org/project/mslm-email-verify-test2/)

UmanShahzad reviewed (approved):
Just some doc fixes then let's just merge and ship this

Harisabdullah commented on CONTRIBUTING.md:
Done

UmanShahzad commented on README.md:
```diff
@@ -13,12 +17,107 @@ Mslm's APIs require an API key. If you don't have one, please read
 ## Installation
-
-...
+To install the main library, use the following command:
```
```suggestion
To install the main library, use the following command:

```

UmanShahzad commented on README.md:
````diff
@@ -13,12 +17,107 @@ Mslm's APIs require an API key. If you don't have one, please read
+```bash
+pip install mslm
+```
+or
+Install `mslm-email-verify` and `mslm-otp` separately.
+```bash
````
```suggestion

Or install specific packages like `mslm-email-verify` and `mslm-otp` separately:

```bash
```
